### PR TITLE
Fixes #726: Implemented two accuracy functions

### LINKED ISSUE #726: 使用SPU实现两个accuracy函数
### Feature Request Type

Build/Install

### Have you searched existing issues?

Yes

### Is your feature request related to a problem?

>此 ISSUE 为 [隐语开源共建计划（SecretFlow Open Source Contribution Plan，简称 SF OSCP)](https://github.com/orgs/secretflow/discussions/1181)任务 ISSUE，欢迎社区开发者参与共建～
若有感兴趣想要认领的任务，但还未报名，辛苦先完成[报名](https://www.wjx.top/vm/QKio0dq.aspx#)进行哈～

## 任务介绍
- **任务名称：** 使用SPU实现两个accuracy函数
- **技术方向：** SPU/SML
- **任务难度：** 热身🌟
- **任务期望时间：** 2-3 周
## 详细要求
- **安全性：** 无reveal
- **功能性：** 需要实现以下几个sklearn中的函数
  a. balanced_accuracy
  b. top_k_accuracy
- **收敛性：** 包含 simulator 跑出的实验数据并且证明收敛/正确性
- **代码规范：** Python 代码需要使用 black+isort 进行格式化（流水线包含代码规范检查卡点）
- **提交说明：** 关联该 isuue 并提交代码至https://github.com/secretflow/spu/tree/main/sml （具体目录位置请与reviewer讨论）
- **特殊说明：** 若某个特性有特殊的限制，如需要 FM128，需要更多 fxp 等需要在注释文档中明确说明
## 能力要求
- 熟悉经典的机器学习算法
- 熟悉 JAX 或 NumPy，可以使用 NumPy 实现算法
## 操作说明
- **可参考：**
  ○ https://scikit-learn.org/stable/modules/generated/sklearn.metrics.balanced_accuracy_score.html
  ○ https://scikit-learn.org/stable/modules/generated/sklearn.metrics.top_k_accuracy_score.html
- **操作指引：** https://www.secretflow.org.cn/docs/spu/latest/en-US/getting_started/tutorials/develop_your_first_mpc_application
- **范例：** https://github.com/secretflow/spu/blob/main/sml/linear_model/simple_sgd.py
- **可参考隐语 MOOC 2期视频:**
  ○ https://www.bilibili.com/video/BV1ba4y1S73z/?spm_id_from=333.788
  ○ https://www.bilibili.com/video/BV1Qe411D7MG/?spm_id_from=333.788
- **贡献指引文档：** https://github.com/secretflow/spu/blob/main/sml/development.md

### PR DESCRIPTION
**Pull request by 1160300918**
# Pull Request

## What problem does this PR solve?

Issue Number: Fixed #726

Implemented two accuracy functions for binary classification and multi-class classification.

### CODE REVIEW

**Comment by 1160300918** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by 1160300918** [comment]
recheck

**Comment by 1160300918** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by 1160300918** [comment]
> Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://github.com/secretflow/.github/blob/main/CLA.md) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.
> 
> I have read the CLA Document and I hereby sign the CLA
> 
> You can retrigger this bot by commenting **recheck** in this Pull Request. Posted by the **CLA Assistant Lite bot**.

recheck

File: sml/metrics/classification/classification.py [hunk 1]
```diff
@@ -16,13 +16,92 @@
...
+def top_k_accuracy_score(
+    y_true, y_score, k=2, normalize=True, sample_weight=None, labels=None
+):
+    """
+    Top-k Accuracy classification score.
+    This metric computes the number of times when the correct label is among
+    the top `k` labels predicted (ranked by predicted scores).
+    """
+
+    # 转换 y_true 和 y_score 为 JAX 数组
```
Comment (rivertalk) [inline sml/metrics/classification/classification.py:75]: No need to comment, code explains itself well here.

File: sml/metrics/classification/classification.py [hunk 2]
```diff
@@ -16,13 +16,92 @@
...
 
 from .auc import binary_clf_curve, binary_roc_auc
 
 
+def confusion_matrix(y_true, y_pred, labels, sample_weight=None, normalize=None):
+    """calculate the confusion matrix"""
+    y_true = jnp.array(y_true)
+    y_pred = jnp.array(y_pred)
+
+    # 获取标签的数量
```
Comment (rivertalk) [inline sml/metrics/classification/classification.py:32]: No need to comment, code explains itself well here.

File: sml/metrics/classification/classification.py [hunk 3]
```diff
@@ -16,13 +16,92 @@
...
+    num_labels = len(labels)
+
+    # 初始化混淆矩阵
+    cm = jnp.zeros((num_labels, num_labels), dtype=jnp.int32)
+
+    # 计算混淆矩阵
```
Comment (rivertalk) [inline sml/metrics/classification/classification.py:38]: Please use English for comments, keep consistent to the project style.

**Comment by 1160300918** [comment]
recheck

**Comment by 1160300918** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by 1160300918** [comment]
recheck

File: sml/metrics/classification/classification.py (hunk 3 above)
Comment (1160300918) [reply sml/metrics/classification/classification.py]: Update code comments with English.

**Comment by 1160300918** [comment]
Can anyone tell me how I can sign the CLA，please?

**Comment by anakinxc** [comment]
Hi @1160300918 

The commits and this pull request are created by different accounts, you have to sign the CLA with both accounts.

**Comment by hesy7** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by 1160300918** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by 1160300918** [comment]
> Hi @1160300918
> 
> The commits and this pull request are created by different accounts, you have to sign the CLA with both accounts.

Thanks. The bot passed, but the CLA Assistant in workflows still failed, should I commit again to activate it？

**Comment by anakinxc** [comment]
> > Hi @1160300918
> > The commits and this pull request are created by different accounts, you have to sign the CLA with both accounts.
> 
> Thanks. The bot passed, but the CLA Assistant in workflows still failed, should I commit again to activate it？

ignore that.

**Comment by 1160300918** [comment]
> > > Hi @1160300918
> > > The commits and this pull request are created by different accounts, you have to sign the CLA with both accounts.
> > 
> > 
> > Thanks. The bot passed, but the CLA Assistant in workflows still failed, should I commit again to activate it？
> 
> ignore that.

OK, is there anything else I should do besides waiting for the code review results?

**Comment by anakinxc** [comment]
> > > > Hi @1160300918
> > > > The commits and this pull request are created by different accounts, you have to sign the CLA with both accounts.
> > > 
> > > 
> > > Thanks. The bot passed, but the CLA Assistant in workflows still failed, should I commit again to activate it？
> > 
> > 
> > ignore that.
> 
> OK, is there anything else I should do besides waiting for the code review results?

just be patient

File: sml/metrics/classification/classification.py [hunk 4]
```diff
@@ -16,13 +16,91 @@
...
+    the top `k` labels predicted (ranked by predicted scores).
+    """
+
+    y_true = jnp.asarray(y_true)
+    y_score = jnp.asarray(y_score)
+
+    if labels is not None:
+        # If labels are provided, make sure y_true and y_score are included in labels
+        labels = jnp.asarray(labels)
+        y_true = jnp.searchsorted(labels, y_true, sorter=jnp.argsort(labels))
```
Comment (deadlywing) [inline sml/metrics/classification/classification.py:81]: Searchsorted is a costly op, you can just skip it.

Then, you should comment in the docs to hint to the user that the y_true should only contain values 0,1,2..., len(labels)-1

File: sml/metrics/classification/classification.py [hunk 5]
```diff
@@ -16,13 +16,91 @@
...
+
+    # Compute the indices of the top k predictions for each sample
+    top_k_indices = jnp.argsort(y_score, axis=1)[:, -k:]
```
Comment (deadlywing) [inline sml/metrics/classification/classification.py:84]: There are some easy cases: Binary case, when k=1,  you just compare to 0.5; k>1, you just return 1 (Indeed, you can always check whether k >= len(labels)).

Comment (deadlywing) [inline sml/metrics/classification/classification.py:84]: In general cases (multi-class case), you can use `jax.lax.top_k`  to replace the full sort

File: sml/metrics/classification/classification.py [hunk 6]
```diff
@@ -16,13 +16,91 @@
...
+
+    # Check if y_true is among the top k predictions
+    y_true_in_top_k = jnp.any(jnp.isin(y_true[:, None], top_k_indices), axis=1)
+
+    # Calculate accuracy
+    correct_predictions = jnp.sum(y_true_in_top_k)
```
Comment (deadlywing) [inline sml/metrics/classification/classification.py:90]: You should think twice about the `sample_weight` and `normalize`.

- if `sample_weight` is not None, and `normalize=False`, `correct_predictions` should also take `sample_weight` into consideration.

Maybe you can refer to [sklearn]( https://github.com/scikit-learn/scikit-learn/blob/8a2d5ffa0/sklearn/metrics/_ranking.py#L2046) about the logic

File: sml/metrics/classification/classification_emul.py [hunk 7]
```diff
@@ -27,15 +27,86 @@
...
+        return balanced_score
+
+    def check(spu_result, sk_result):
+        np.testing.assert_allclose(spu_result, sk_result, rtol=1, atol=1e-5)
+
+    # Test binary
+    y_true = jnp.array([0, 1, 1, 0, 1, 1])
+    y_pred = jnp.array([0, 0, 1, 0, 1, 1])
+    labels = jnp.array([0, 1])
+    spu_result = emulator.run(proc)(y_true, y_pred, labels)
```
Comment (deadlywing) [inline sml/metrics/classification/classification_emul.py:58]: you should seal the input tensors before feeding them into the emulator, you can refer to other `emul_` funcs

File: sml/metrics/classification/classification_emul.py [hunk 8]
```diff
@@ -27,15 +27,86 @@
...
+    y_score = jnp.array(
+        [
+            [0.8, 0.1, 0.1],
+            [0.3, 0.4, 0.3],
+            [0.1, 0.1, 0.8],
+            [0.2, 0.2, 0.6],
+            [0.7, 0.2, 0.1],
+        ]
+    )
+    spu_result = emulator.run(proc, static_argnums=(2, 3))(
```
Comment (deadlywing) [inline sml/metrics/classification/classification_emul.py:103]: Likewise, should seal the data beforehand

File: sml/metrics/classification/classification_test.py [hunk 9]
```diff
@@ -33,16 +33,100 @@
...
+        y_score = jnp.array(
+            [
+                [0.8, 0.1, 0.1],
+                [0.3, 0.4, 0.3],
+                [0.1, 0.1, 0.8],
+                [0.2, 0.2, 0.6],
+                [0.7, 0.2, 0.1],
+            ]
+        )
+        spu_result = spsim.sim_jax(sim, proc, static_argnums=(2, 3))(
```
Comment (deadlywing) [inline sml/metrics/classification/classification_test.py:124]: Test more please, at least you should test `sample_weight`, and `normalize`.
And you should check for some larger data, maybe a size of 1000 is the least requirement.

File: sml/metrics/classification/classification_test.py [hunk 10]
```diff
@@ -33,16 +33,100 @@
...
+            return top_k_score
+
+        def sklearn_proc(y_true, y_pred, k, labels):
+            top_k_score = metrics.top_k_accuracy_score(
+                y_true, y_pred, k=k, labels=labels
+            )
+            return top_k_score
+
+        def check(spu_result, sk_result):
+            np.testing.assert_allclose(spu_result, sk_result, rtol=1, atol=1e-5)
```
Comment (deadlywing) [inline sml/metrics/classification/classification_test.py:111]: `rtol` and `atol` can be set to `1e-3`

File: sml/metrics/classification/classification_test.py [hunk 11]
```diff
@@ -33,16 +33,100 @@
...
+        def proc(y_true: jnp.ndarray, y_pred: jnp.ndarray, labels: jnp.ndarray):
+            balanced_score = balanced_accuracy_score(y_true, y_pred, labels)
+            return balanced_score
+
+        def sklearn_proc(y_true, y_pred):
+            balanced_score = metrics.balanced_accuracy_score(y_true, y_pred)
+            return balanced_score
+
+        def check(spu_result, sk_result):
+            np.testing.assert_allclose(spu_result, sk_result, rtol=1, atol=1e-5)
```
Comment (deadlywing) [inline sml/metrics/classification/classification_test.py:63]: `rtol` and `atol` can be set to `1e-3`

File: sml/metrics/classification/classification_test.py [hunk 12]
```diff
@@ -33,16 +33,100 @@
...
     roc_auc_score,
+    top_k_accuracy_score,
 )
 
 
 class UnitTests(unittest.TestCase):
+
+    def test_balanced_accuracy(self):
+        sim = spsim.Simulator.simple(
+            3, spu_pb2.ProtocolKind.ABY3, spu_pb2.FieldType.FM128
```
Comment (deadlywing) [inline sml/metrics/classification/classification_test.py:51]: `FM64` is enough

File: sml/metrics/classification/classification_test.py [hunk 13]
```diff
@@ -33,16 +33,100 @@
...
+        y_true = jnp.array([0, 1, 1, 0, 2, 1])
+        y_pred = jnp.array([0, 0, 1, 0, 2, 1])
+        labels = jnp.array([0, 1, 2])
+        spu_result = spsim.sim_jax(sim, proc)(y_true, y_pred, labels)
+        sk_result = sklearn_proc(y_true, y_pred)
+        check(spu_result, sk_result)
+
+    def test_top_k_accuracy(self):
+        sim = spsim.Simulator.simple(
+            3, spu_pb2.ProtocolKind.ABY3, spu_pb2.FieldType.FM128
```
Comment (deadlywing) [inline sml/metrics/classification/classification_test.py:83]: `FM64` is enough

File: sml/metrics/classification/classification_test.py [hunk 14]
```diff
@@ -33,16 +33,100 @@
...
+            return balanced_score
+
+        def check(spu_result, sk_result):
+            np.testing.assert_allclose(spu_result, sk_result, rtol=1, atol=1e-5)
+
+        # Test binary
+        y_true = jnp.array([0, 1, 1, 0, 1, 1])
+        y_pred = jnp.array([0, 0, 1, 0, 1, 1])
+        labels = jnp.array([0, 1])
+        spu_result = spsim.sim_jax(sim, proc)(y_true, y_pred, labels)
```
Comment (deadlywing) [inline sml/metrics/classification/classification_test.py:69]: Test the param `sample_weight` and `adjusted`;
Test larger datasets, please (maybe ~1000 samples are enough).
